File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureFilePermissions.java
package co.ostorlab.insecure_app.bugs.calls;

import android.content.Context;
import java.io.File;
import co.ostorlab.insecure_app.BugRule;


public final class InsecureFilePermissions extends BugRule {

    private static final String TAG = "RULE";

    @Override
    public String getDescription() {
        return "Use of insecure file permissions";
    }

    @Override
    public void run() throws Exception {
        String filename = "test_filename";
        openFileOutputWorldReadable(filename);
        openFileOutputWorldWritable(filename);
        setReadableAll(filename);
        setWritableAll(filename);
    }

    private void openFileOutputWorldReadable(String filename) throws Exception {
        getContext().openFileOutput(filename, Context.MODE_WORLD_READABLE);
    }

    private void openFileOutputWorldWritable(String filename) throws Exception {
        getContext().openFileOutput(filename, Context.MODE_WORLD_WRITEABLE);
    }

    private void setReadableAll(String filename) throws Exception {
        // see https://stackoverflow.com/questions/13856757/android-alternative-to-the-deprecated-context-mode-world-readable/**/
        getContext().openFileOutput(filename, Context.MODE_PRIVATE);
        String path = getContext().getFilesDir().getAbsolutePath() + filename;
        File f = new File(path);
        f.setReadable(true, false);
    }

    private void setWritableAll(String filename) throws Exception {
        getContext().openFileOutput(filename, Context.MODE_PRIVATE);
        String path = getContext().getFilesDir().getAbsolutePath() + filename;
        File f = new File(path);
        f.setWritable(true, false);
    }

}

File: ostorlab_insecure_flutter_app/lib/bugs/insecure_file_permissions.dart
import 'dart:io';
import 'package:ostorlab_insecure_flutter_app/bug_rule.dart';

/// A [BugRule] implementation that triggers the use of insecure file permissions.
class InsecureFilePermissions extends BugRule {
  /// The tag used to identify instances of this rule.
  static const String _tag = "InsecureFilePermissions";

  /// Returns a description of this [BugRule] implementation.
  @override
  String getDescription() {
    return "Use of insecure file permissions";
  }

  /// Trigger the bug rule.
  @override
  Future<void> run() async {
    String filename = "test_filename";
    await openFileOutputWorldReadable(filename);
    await openFileOutputWorldWritable(filename);
    await setReadableAll(filename);
    await setWritableAll(filename);
  }

  /// Writes to a file with world-readable permission.
  Future<void> openFileOutputWorldReadable(String filename) async {
    final documentsDir = '/sdcard/Documents';
    final file = await File('${documentsDir}/$filename').create();
    await file.writeAsString('world readable');
    await _setFilePermissions(file.path, '444');
  }

  /// Writes to a file with world-writable permission.
  Future<void> openFileOutputWorldWritable(String filename) async {
    final documentsDir = '/sdcard/Documents';
    final file = await File('${documentsDir}/$filename').create();
    await file.writeAsString('world writable');
    await _setFilePermissions(file.path, '666');
  }

  /// Sets a file readable to all users.
  Future<void> setReadableAll(String filename) async {
    final documentsDir = '/sdcard/Documents';
    final file = await File('${documentsDir}/$filename').create();
    await file.writeAsString('readable all');
    await _setFilePermissions(file.path, '444');
  }

  /// Sets a file writable to all users.
  Future<void> setWritableAll(String filename) async {
    final documentsDir = '/sdcard/Documents';
    final file = await File('${documentsDir}/$filename').create();
    await file.writeAsString('writable all');
    await _setFilePermissions(file.path, '666');
  }

  /// Sets the permissions of a file using the chmod command.
  Future<void> _setFilePermissions(String filePath, String permission) async {
    await Process.run('chmod', [permission, filePath]);
  }
}
